node/pkg/p2p/run_params.go

```diff
@@ -0,0 +1,220 @@
+		queryResponseReadC    <-chan *query.QueryResponsePublication
+		ccqBootstrapPeers     string
+		ccqPort               uint
+		ccqAllowedPeers       string
+	}
+
+	// RunOpt is used to specify optional parameters.
+	RunOpt func(p *RunParams) error
+)
+
+// NewRunParams is used to create the `RunParams` which gets passed into `p2p.Run()`. It takes the required parameters,
+// plus any desired optional ones, which can be set using the various `With` functions defined below.
+func NewRunParams(
+	bootstrapPeers string,
+	networkID string,
+	priv crypto.PrivKey,
+	gst *common.GuardianSetState,
+	rootCtxCancel context.CancelFunc,
+	opts ...RunOpt,
+) (*RunParams, error) {
+	p := &RunParams{
+		bootstrapPeers: bootstrapPeers,
+		networkID:      networkID,
+		priv:           priv,
+		gst:            gst,
+		rootCtxCancel:  rootCtxCancel,
+	}
+
+	for _, opt := range opts {
+		err := opt(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.verify(); err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// WithSignedObservationListener is used to set the channel to receive `SignedObservation“ messages.
+func WithSignedObservationListener(obsvC chan<- *common.MsgWithTimeStamp[gossipv1.SignedObservation]) RunOpt {
+	return func(p *RunParams) error {
+		p.obsvC = obsvC
+		return nil
+	}
+}
+
+// WithSignedVAAListener is used to set the channel to receive `SignedVAAWithQuorum messages.
+func WithSignedVAAListener(signedInC chan<- *gossipv1.SignedVAAWithQuorum) RunOpt {
+	return func(p *RunParams) error {
+		p.signedInC = signedInC
+		return nil
+	}
+}
+
+// WithObservationRequestListener is used to set the channel to receive `ObservationRequest messages.
+func WithObservationRequestListener(obsvReqC chan<- *gossipv1.ObservationRequest) RunOpt {
+	return func(p *RunParams) error {
+		p.obsvReqC = obsvReqC
+		return nil
+	}
+}
+
+// WithChainGovernorConfigListener is used to set the channel to receive `SignedChainGovernorConfig messages.
+func WithChainGovernorConfigListener(signedGovCfg chan *gossipv1.SignedChainGovernorConfig) RunOpt {
+	return func(p *RunParams) error {
+		p.signedGovCfg = signedGovCfg
+		return nil
+	}
+}
+
+// WithChainGovernorStatusListener is used to set the channel to receive `SignedChainGovernorStatus messages.
+func WithChainGovernorStatusListener(signedGovSt chan *gossipv1.SignedChainGovernorStatus) RunOpt {
+	return func(p *RunParams) error {
+		p.signedGovSt = signedGovSt
+		return nil
+	}
+}
+
+// WithDisableHeartbeatVerify is used to set disableHeartbeatVerify.
+func WithDisableHeartbeatVerify(disableHeartbeatVerify bool) RunOpt {
+	return func(p *RunParams) error {
+		p.disableHeartbeatVerify = disableHeartbeatVerify
+		return nil
+	}
+}
+
+// WithGuardianOptions is used to set options that are only meaningful to the guardian.
+func WithGuardianOptions(
+	nodeName string,
+	gk *ecdsa.PrivateKey,
+	obsvC chan<- *common.MsgWithTimeStamp[gossipv1.SignedObservation],
```

Comment:
Isn't this one set with WithSignedObservationListener?

Reply:
Is this so that you don't have to call all the functions and just pass all the values in?

Reply:
Yes, but that is meant to be used by clients. I wanted `WithGuardianOptions` to take everything that the guardian requires.